# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/gtk/DMABufSurface.cpp
# Commit: a8ce70eff417
# Full Hash: a8ce70eff41744ce26905ea160b21b63516c461b
# Author: stransky <stransky@redhat.com>
# Date: 2025-01-08 04:42:02
# Regressor Bug: 1940037
# File Overlap Count: 1
# Description:
#   Bug 1940037 [Wayland] Use exact FOURCC format for DMABuf surfaces r=emilio
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D233241
# ==============================================================================

diff -r 20a5670ea914 -r a8ce70eff417 widget/gtk/DMABufSurface.cpp
--- a/widget/gtk/DMABufSurface.cpp	Tue Jan 07 11:58:50 2025 +0000
+++ b/widget/gtk/DMABufSurface.cpp	Tue Jan 07 11:58:51 2025 +0000
@@ -506,14 +506,14 @@
     // Requested DRM format is not supported.
     return false;
   }
-  mDrmFormats[0] = mGmbFormat->mFormat;
+  mFOURCCFormat = mGmbFormat->mFormat;
 
   bool useModifiers = (aDMABufSurfaceFlags & DMABUF_USE_MODIFIERS) &&
                       !mGmbFormat->mModifiers.IsEmpty();
   if (useModifiers) {
     LOGDMABUF(("    Creating with modifiers\n"));
     mGbmBufferObject[0] = GbmLib::CreateWithModifiers(
-        GetDMABufDevice()->GetGbmDevice(), mWidth, mHeight, mDrmFormats[0],
+        GetDMABufDevice()->GetGbmDevice(), mWidth, mHeight, mFOURCCFormat,
         mGmbFormat->mModifiers.Elements(), mGmbFormat->mModifiers.Length());
     if (mGbmBufferObject[0]) {
       mBufferModifiers[0] = GbmLib::GetModifier(mGbmBufferObject[0]);
@@ -525,7 +525,7 @@
     mGbmBufferFlags = GBM_BO_USE_LINEAR;
     mGbmBufferObject[0] =
         GbmLib::Create(GetDMABufDevice()->GetGbmDevice(), mWidth, mHeight,
-                       mDrmFormats[0], mGbmBufferFlags);
+                       mFOURCCFormat, mGbmBufferFlags);
     mBufferModifiers[0] = DRM_FORMAT_MOD_INVALID;
   }
 
@@ -604,7 +604,7 @@
   }
 
   LOGDMABUF(("  imported size %d x %d format %x planes %d modifiers %" PRIx64,
-             mWidth, mHeight, mDrmFormats[0], mBufferPlaneCount,
+             mWidth, mHeight, mFOURCCFormat, mBufferPlaneCount,
              mBufferModifiers[0]));
   return true;
 }
@@ -620,7 +620,7 @@
   mHeight = aHeight;
   mBufferModifiers[0] = aDMABufInfo.modifier;
   mGmbFormat = GetDMABufDevice()->GetGbmFormat(true);
-  mDrmFormats[0] = mGmbFormat->mFormat;
+  mFOURCCFormat = mGmbFormat->mFormat;
   mBufferPlaneCount = aDMABufInfo.plane_count;
 
   RefPtr<gfx::FileHandleWrapper> fd = std::move(aFd);
@@ -632,7 +632,7 @@
   }
 
   LOGDMABUF(("  imported size %d x %d format %x planes %d modifiers %" PRIx64,
-             mWidth, mHeight, mDrmFormats[0], mBufferPlaneCount,
+             mWidth, mHeight, mFOURCCFormat, mBufferPlaneCount,
              mBufferModifiers[0]));
   return true;
 }
@@ -641,10 +641,9 @@
     const SurfaceDescriptor& aDesc) {
   const SurfaceDescriptorDMABuf& desc = aDesc.get_SurfaceDescriptorDMABuf();
 
+  mFOURCCFormat = desc.fourccFormat();
   mWidth = desc.width()[0];
   mHeight = desc.height()[0];
-  mBufferModifiers[0] = desc.modifier()[0];
-  mDrmFormats[0] = desc.format()[0];
   mBufferPlaneCount = desc.fds().Length();
   mGbmBufferFlags = desc.flags();
   MOZ_RELEASE_ASSERT(mBufferPlaneCount <= DMABUF_BUFFER_PLANES);
@@ -658,6 +657,8 @@
     mDmabufFds[i] = desc.fds()[i];
     mStrides[i] = desc.strides()[i];
     mOffsets[i] = desc.offsets()[i];
+    mDrmFormats[i] = desc.format()[i];
+    mBufferModifiers[i] = desc.modifier()[i];
   }
 
   if (desc.fence().Length() > 0) {
@@ -673,7 +674,7 @@
   }
 
   LOGDMABUF(("  imported size %d x %d format %x planes %d", mWidth, mHeight,
-             mDrmFormats[0], mBufferPlaneCount));
+             mFOURCCFormat, mBufferPlaneCount));
   return true;
 }
 
@@ -703,12 +704,12 @@
 
   width.AppendElement(mWidth);
   height.AppendElement(mHeight);
-  format.AppendElement(mDrmFormats[0]);
-  modifiers.AppendElement(mBufferModifiers[0]);
   for (int i = 0; i < mBufferPlaneCount; i++) {
     fds.AppendElement(WrapNotNull(mDmabufFds[i]));
     strides.AppendElement(mStrides[i]);
     offsets.AppendElement(mOffsets[i]);
+    format.AppendElement(mDrmFormats[i]);
+    modifiers.AppendElement(mBufferModifiers[i]);
   }
 
   CloseFileDescriptors(lockFD);
@@ -722,9 +723,9 @@
   }
 
   aOutDescriptor = SurfaceDescriptorDMABuf(
-      mSurfaceType, modifiers, mGbmBufferFlags, fds, width, height, width,
-      height, format, strides, offsets, GetYUVColorSpace(), mColorRange,
-      mozilla::gfx::ColorSpace2::UNKNOWN,
+      mSurfaceType, mFOURCCFormat, modifiers, mGbmBufferFlags, fds, width,
+      height, width, height, format, strides, offsets, GetYUVColorSpace(),
+      mColorRange, mozilla::gfx::ColorSpace2::UNKNOWN,
       mozilla::gfx::TransferFunction::Default, fenceFDs, mUID, refCountFDs,
       /* semaphoreFd */ nullptr);
   return true;
@@ -740,7 +741,7 @@
   attribs.AppendElement(LOCAL_EGL_HEIGHT);
   attribs.AppendElement(mHeight);
   attribs.AppendElement(LOCAL_EGL_LINUX_DRM_FOURCC_EXT);
-  attribs.AppendElement(mDrmFormats[0]);
+  attribs.AppendElement(mFOURCCFormat);
 #define ADD_PLANE_ATTRIBS(plane_idx)                                        \
   {                                                                         \
     attribs.AppendElement(LOCAL_EGL_DMA_BUF_PLANE##plane_idx##_FD_EXT);     \
@@ -1204,6 +1205,7 @@
     return false;
   }
   mSurfaceType = SURFACE_YUV;
+  mFOURCCFormat = aDesc.fourcc;
   mBufferPlaneCount = aDesc.num_layers;
 
   for (unsigned int i = 0; i < aDesc.num_layers; i++) {
@@ -1377,7 +1379,8 @@
 bool DMABufSurfaceYUV::UpdateYUVData(void** aPixelData, int* aLineSizes) {
   LOGDMABUF(("DMABufSurfaceYUV::UpdateYUVData() UID %d", mUID));
 
-  if (mSurfaceType != SURFACE_YUV || mBufferPlaneCount != 3) {
+  if (mSurfaceType != SURFACE_YUV || mBufferPlaneCount != 3 ||
+      mFOURCCFormat != VA_FOURCC_YV12) {
     LOGDMABUF(("    UpdateYUVData can upload YUV420 surface type only!"));
     return false;
   }
@@ -1415,6 +1418,7 @@
              aHeight));
 
   mSurfaceType = SURFACE_YUV;
+  mFOURCCFormat = VA_FOURCC_YV12;
   mBufferPlaneCount = 3;
 
   if (!CreateLinearYUVPlane(0, aWidth, aHeight, GBM_FORMAT_R8)) {
@@ -1440,6 +1444,7 @@
     const SurfaceDescriptorDMABuf& aDesc) {
   mBufferPlaneCount = aDesc.fds().Length();
   mSurfaceType = SURFACE_YUV;
+  mFOURCCFormat = aDesc.fourccFormat();
   mColorSpace = aDesc.yUVColorSpace();
   mColorRange = aDesc.colorRange();
   mColorPrimaries = aDesc.colorPrimaries();
@@ -1519,8 +1524,8 @@
   }
 
   aOutDescriptor = SurfaceDescriptorDMABuf(
-      mSurfaceType, modifiers, 0, fds, width, height, widthBytes, heightBytes,
-      format, strides, offsets, GetYUVColorSpace(), mColorRange,
+      mSurfaceType, mFOURCCFormat, modifiers, 0, fds, width, height, widthBytes,
+      heightBytes, format, strides, offsets, GetYUVColorSpace(), mColorRange,
       mColorPrimaries, mTransferFunction, fenceFDs, mUID, refCountFDs,
       /* semaphoreFd */ nullptr);
   return true;